File: mns-reactor/src/test/java/cz/fb/manaus/reactor/charge/ChargeGrowthForecasterTest.java
package cz.fb.manaus.reactor.charge;

import cz.fb.manaus.core.model.Bet;
import cz.fb.manaus.core.model.MarketSnapshot;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.Side;
import cz.fb.manaus.core.provider.ExchangeProvider;
import cz.fb.manaus.core.test.AbstractLocalTestCase;
import cz.fb.manaus.core.test.CoreTestFactory;
import cz.fb.manaus.reactor.ReactorTestFactory;
import cz.fb.manaus.reactor.betting.AmountAdviser;
import cz.fb.manaus.reactor.price.FairnessPolynomialCalculator;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.Date;
import java.util.LinkedList;
import java.util.List;
import java.util.Optional;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;


public class ChargeGrowthForecasterTest extends AbstractLocalTestCase {

    @Autowired
    private ChargeGrowthForecaster forecaster;
    @Autowired
    private ReactorTestFactory factory;
    @Autowired
    private FairnessPolynomialCalculator calculator;
    @Autowired
    private AmountAdviser adviser;
    @Autowired
    private ExchangeProvider provider;

    @Test
    public void testForecast() {
        var market = factory.createMarket(0.05, List.of(0.5, 0.3, 0.2));
        var currentBets = new LinkedList<Bet>();
        var marketSnapshot = MarketSnapshot.from(market, currentBets, Optional.empty());
        var fairness = calculator.getFairness(market);
        var forecast = forecaster.getForecast(CoreTestFactory.DRAW, Side.BACK, marketSnapshot, fairness);
        assertTrue(forecast.getAsDouble() > 1);
        var betAmount = adviser.getAmount();
        currentBets.add(new Bet("1", CoreTestFactory.MARKET_ID, CoreTestFactory.DRAW,
                new Price(3d, betAmount, Side.LAY), new Date(), betAmount));
        forecast = forecaster.getForecast(CoreTestFactory.DRAW, Side.BACK, marketSnapshot, fairness);
        assertFalse(forecast.getAsDouble() > 1);

        forecast = forecaster.getForecast(CoreTestFactory.HOME, Side.BACK, marketSnapshot, fairness);
        assertTrue(forecast.getAsDouble() > 1);

        forecast = forecaster.getForecast(CoreTestFactory.HOME, Side.LAY, marketSnapshot, fairness);
        assertFalse(forecast.getAsDouble() > 1);
    }
}

File: mns-reactor/src/test/java/cz/fb/manaus/reactor/charge/ChargeGrowthForecasterTest.kt
package cz.fb.manaus.reactor.charge

import cz.fb.manaus.core.model.*
import cz.fb.manaus.core.test.AbstractTestCase5
import cz.fb.manaus.reactor.PricesTestFactory
import cz.fb.manaus.reactor.betting.AmountAdviser
import cz.fb.manaus.reactor.price.FairnessPolynomialCalculator
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import java.time.Instant
import kotlin.test.assertFalse
import kotlin.test.assertTrue


class ChargeGrowthForecasterTest : AbstractTestCase5() {

    @Autowired
    private lateinit var forecaster: ChargeGrowthForecaster

    @Autowired
    private lateinit var adviser: AmountAdviser

    @Test
    fun forecast() {
        val marketPrices = PricesTestFactory.newMarketPrices(0.05, listOf(0.5, 0.3, 0.2))
        val currentBets = mutableListOf<TrackedBet>()
        val snapshot = MarketSnapshot(marketPrices, market, currentBets)
        val fairness = FairnessPolynomialCalculator.getFairness(marketPrices)
        val commission = bfProvider.commission
        val homeBack = SideSelection(Side.BACK, SEL_HOME)
        var forecast = forecaster.getForecast(homeBack, snapshot, fairness, commission)
        assertTrue(forecast!! > 1)
        val betAmount = adviser.amount
        currentBets.add(
                Bet(
                        "1", "1", SEL_DRAW,
                        Price(3.0, betAmount, Side.LAY),
                        Instant.now(),
                        betAmount
                ).asTracked
        )
        val drawBack = SideSelection(Side.BACK, SEL_DRAW)
        forecast = forecaster.getForecast(drawBack, snapshot, fairness, commission)
        assertFalse(forecast!! > 1)

        forecast = forecaster.getForecast(homeBack, snapshot, fairness, commission)
        assertTrue(forecast!! > 1)

        val homeLay = SideSelection(Side.LAY, SEL_HOME)
        forecast = forecaster.getForecast(homeLay, snapshot, fairness, commission)
        assertFalse(forecast!! > 1)
    }
}